external-crates/move/crates/move-compiler/src/typing/visitor.rs

```diff
@@ -163,24 +241,149 @@ pub trait TypingVisitorContext {
+                .map(|(_, _, ty)| ty)
+                .for_each(|ty| self.visit_type(ty));
+            self.visit_type(&mut fdef.signature.return_type);
+        }
         if let T::FunctionBody_::Defined(seq) = &mut fdef.body.value {
             self.visit_seq(seq);
         }
         self.pop_warning_filter_scope();
     }
 
-    fn visit_seq(&mut self, (_, seq): &mut T::Sequence) {
+    // -- TYPES --
+
+    fn visit_type_custom(&mut self, _ty: &mut N::Type) -> bool {
+        false
+    }
+
+    /// Visit a type, including recursively. Note that this may be called manually even if
+    /// `VISIT_TYPES` is set to `false`.
+    #[growing_stack]
+    fn visit_type(&mut self, ty: &mut N::Type) {
+        if self.visit_type_custom(ty) {
+            return;
+        }
+        match &mut ty.value {
+            N::Type_::Unit => (),
+            N::Type_::Ref(_, inner) => self.visit_type(inner),
+            N::Type_::Param(_) => (),
+            N::Type_::Apply(_, _, args) => args.iter_mut().for_each(|ty| self.visit_type(ty)),
+            N::Type_::Fun(args, ret) => {
+                args.iter_mut().for_each(|ty| self.visit_type(ty));
+                self.visit_type(ret);
+            }
+            N::Type_::Var(_) => (),
+            N::Type_::Anything => (),
+            N::Type_::UnresolvedError => (),
+        }
+    }
+
+    // -- USE FUNS --
+
+    fn visit_use_funs_custom(&mut self, _use_funs: &mut N::UseFuns) -> bool {
+        false
+    }
+
+    fn visit_use_funs(&mut self, use_funs: &mut N::UseFuns) {
+        let _ = self.visit_use_funs_custom(use_funs);
+        // Nothing to traverse in the other case
+    }
+
+    // -- SEQUENCES AND EXPRESSIONS --
+
+    fn visit_seq(&mut self, (use_funs, seq): &mut T::Sequence) {
+        if Self::VISIT_USE_FUNS {
+            self.visit_use_funs(use_funs);
+        }
         for s in seq {
             self.visit_seq_item(s);
         }
     }
 
-    fn visit_seq_item(&mut self, sp!(_, seq_item): &mut T::SequenceItem) {
+    /// Custom visit for a sequence item. It will skip `visit_seq_item` if `visit_seq_item_custom`
+    /// returns true.
+    fn visit_seq_item_custom(&mut self, _seq_item: &mut T::SequenceItem) -> bool {
+        false
+    }
+
+    fn visit_seq_item(&mut self, seq_item: &mut T::SequenceItem) {
         use T::SequenceItem_ as SI;
-        match seq_item {
+        if self.visit_seq_item_custom(seq_item) {
+            return;
+        }
+        match &mut seq_item.value {
             SI::Seq(e) => self.visit_exp(e),
+            SI::Declare(lvalues) if Self::VISIT_LVALUES => {
+                self.visit_lvalue_list(&LValueKind::Bind, lvalues);
+            }
             SI::Declare(_) => (),
-            SI::Bind(_, _, e) => self.visit_exp(e),
+            SI::Bind(lvalues, ty_ann, e) => {
+                // visit the RHS first to better match control flow
+                self.visit_exp(e);
+                if Self::VISIT_LVALUES {
```

Comment:
Should we put these sort of gates inside of the respective functions, e.g. `self.visit_lvalue_list` checks `if !Self::VISIT_LVALUES { return }`

Will Rust still be able to compile these calls away?

Reply:
Rust might erase the calls if they are gated and inlining catches / erases them. That doesn't happen in debug mode, but may in release mode. My other motivation for gating them externally is that someone can still call, e.g., `visit_type` in a non-`VISIT_TYPES` visitor if they want to in just a few locations, allowing for a sliver more flexibility.